### PR TITLE
Fixes #3265 where invalid env keys would cause an unhandled exceptio…

### LINKED ISSUE #3265: Unhandled Error when unknown env keys are used
Consider the config below.  Due to a mistake a env key "require" was added.  This causes eslint to blow up.  Instead, the env keys should be validated against known values.

```
{
  "parser": "babel-eslint",
  "env":{
    "browser":true,
    "es6":true,
    "mocha":true,
    "amd": true,
    "require": true
  },
  "ecmaFeatures":{
  },
  "plugins": [
  ],
  "rules": {
  }
}
```

### PR DESCRIPTION
**Pull request by raybooysen**
…n.  Now the env keys are validated against the known keys

### CODE REVIEW

**Comment by nzakas** [comment]
Thanks for the PR. Please take a look at the issue for discussion on where this change should be made. Also note that we require tests for all changes.

**Comment by raybooysen** [comment]
PR ready for review.

**Comment by ilyavolodin** [comment]
@btmills Could you review please?

**Comment by btmills** [comment]
Warning looks good. One concern - this won't warn if config.env has a primitive value. Right now it only runs if config.env is an object or an array.

**Comment by raybooysen** [comment]
Shall I add the check?

**Comment by nzakas** [comment]
Yeah, we should make sure to check as much as possible. Basically:

1. If `env` is an array, throw an error saying it should be a non-array object.
2. If `env` is a non-array object, throw an error if any property name is unknown.

**Comment by nzakas** [comment]
@raybooysen are you going to continue working on this?

**Comment by raybooysen** [comment]
Yes I am, apologies for the delay, I've had a busy week at work.

**Comment by nzakas** [comment]
No problem - just give us an ETA so we can track where this is. If you'd like someone else to take over, just let us know.

**Comment by raybooysen** [comment]
I have an easy day at work tomorrow so will do it then. Apologies for the
dealy5

**Comment by raybooysen** [comment]
@nzakas code is ready, can you review?

File: lib/config-validator.js [hunk 1]
```diff
@@ -95,17 +100,52 @@ function validateRuleOptions(id, options, source) {
...
 /**
+ * Validates an environment object
+ * @param {object} environment The environment config object to validate.
+ * @param {string} source The location to report with any errors.
+ * @returns {void}
+ */
+function validateEnvironment(environment, source) {
+
+    // undefined environments are ok
+    if (environment === "undefined" || environment == null) {
```
Comment (nzakas) [inline lib/config-validator.js]: Are you meaning to do this?

```
if (!environment)
```

or 

```js
if (environment === undefined || environment == null)
```

I'm just not clear on how `environment` could end up equal to the string `"undefined"`.

Comment (raybooysen) [inline lib/config-validator.js]: Yes, typo on my part.  I'll fix

**Comment by raybooysen** [comment]
@nzakas fixed

**Comment by nzakas** [comment]
Thanks! @btmills any last thoughts?

**Comment by btmills** [comment]
Implementation and tests look good! @nzakas how strict are you on commit message length? Something like "Fix: Validate config environments (fixes #xxxx)" would fit in the first line better, but if that's not a concern then this is good to merge.

**Comment by nzakas** [comment]
Ooh yeah, a shorter commit message would be great. Right now, our change log will cut off at about 75 chars, so we would miss the issue number.

**Comment by raybooysen** [comment]
@nzakas @btmills commit message changed

**Pull request merged**